crates/holochain/src/core/workflow/app_validation_workflow.rs

```diff
@@ -485,18 +497,55 @@ fn zome_id_to_zome(zome_id: ZomeId, dna_def: &DnaDef) -> AppValidationResult<Zom
         .into())
 }
 
-fn run_validation_callback_inner(
+#[async_recursion::async_recursion]
+async fn run_validation_callback_inner<R>(
     invocation: ValidateInvocation,
-    ribosome: &impl RibosomeT,
-    workspace_lock: HostFnWorkspaceRead,
+    ribosome: &R,
+    workspace_read: HostFnWorkspaceRead,
     network: HolochainP2pDna,
-) -> AppValidationResult<Outcome> {
-    let validate: ValidateResult =
-        ribosome.run_validate(ValidateHostAccess::new(workspace_lock, network), invocation)?;
-    match validate {
+    fetched_deps: &mut HashSet<AnyDhtHash>,
+) -> AppValidationResult<Outcome>
+where
+    R: RibosomeT + Sync,
+{
+    let validate_result = ribosome.run_validate(
+        ValidateHostAccess::new(workspace_read.clone(), network.clone()),
+        invocation.clone(),
+    )?;
+    match validate_result {
         ValidateResult::Valid => Ok(Outcome::Accepted),
         ValidateResult::Invalid(reason) => Ok(Outcome::Rejected(reason)),
-        ValidateResult::UnresolvedDependencies(hashes) => Ok(Outcome::AwaitingDeps(hashes)),
+        ValidateResult::UnresolvedDependencies(hashes) => {
+            if !hashes.iter().any(|hash| !fetched_deps.contains(hash)) {
```

Comment:
```suggestion
            if hashes.iter().all(|hash| fetched_deps.contains(hash)) {
```
A little easier to parse?

Reply:
Also might be worth a comment here stating this is the base case that ends the recursion. For some devs (like me) it doesn't come naturally to parse recursive loops.
I also wonder what is the maximum recursive depth? Could this panic? Maybe it would be simpler to just use a loop?

Reply:
@freesig i had a loop earlier and it certainly didn't make it look clearer to me, but i can easily limit the call depth to something like 10